egs/wsj/s5/steps/nnet3/multilingual/allocate_multilingual_examples.py

```diff
@@ -150,46 +146,27 @@ def process_multilingual_egs(args):
 
     # Total number of egs in all languages
     tot_num_egs = sum(lang_to_num_examples[i] for i in range(num_langs))
-    if args.num_archives is not None:
-        num_archives = args.num_archives
-    else:
-        num_archives = tot_num_egs // args.samples_per_iter + 1
-
+    num_archives = args.num_archives
 
     with open("{0}/info/{1}num_archives".format(args.egs_dir, args.egs_prefix), "w") as fh:
         print("{0}".format(num_archives), file=fh)
 
-    egs_per_archive = tot_num_egs // num_archives
-    avg_block_size = args.block_size
-
-    # This is the number of robin rounds we do for generating each output scp (in
-    # each round we read one block from each input lang and write it to the current
-    # output scp file; please see the main 'for' loop below):
-    num_rounds = egs_per_archive // (avg_block_size * num_langs)
-
-    # The block size for each lang is calculated based on its size (and average block size).
-    # Note that we can't use the same block size for all input languages; otherwise
-    # we might run out of egs for some of the languages.
-    lang_to_block_size = [int(lang_to_num_examples[i] / tot_num_egs * avg_block_size * num_langs) + 1
-                          for i in range(num_langs)]
-
-    # Due to integer roundings this could be slightly different from 'avg_block_size * num_langs'
-    sum_of_block_sizes = sum(lang_to_block_size)
-    logger.info("Number of egs in each output scp file (except the last one) "
-                "is {}".format(num_rounds * sum_of_block_sizes))
-
-    for lang in range(num_langs):
-        logger.info("block size for lang {} is {}".format(lang, lang_to_block_size[lang]))
+    logger.info("There are a total of {} examples in the input scp "
+                "files.".format(tot_num_egs))
+    logger.info("Number of blocks in each output archive will be approximately "
+                "{}, and block-size is {}.".format(int(round(tot_num_egs / num_archives / args.block_size)),
+                                                   args.block_size))
 
     in_scp_file_handles = [open(scp_lists[lang], 'r') for lang in range(num_langs)]
 
-
-    # For each output scp file, read the examples in a round-robin fashion
-    # from the inputs scp files. Each time exactly one block is read, where the size of the
-    # block is proportional to the total number of egs in the corresponding lang;
-    # so that there is a bit of each lang in each generated scp file.
+    num_remaining_egs = tot_num_egs
+    lang_to_num_remaining_egs = [n for n in lang_to_num_examples]
     for archive_index in range(num_archives):
-        logger.info("Generating archive {}...".format(archive_index))
+        num_remaining_archives = num_archives - archive_index
+        num_remaining_blocks = num_remaining_egs / args.block_size
+
+        num_blocks_per_this_archive = int(round(num_remaining_blocks / num_remaining_archives))
```

Comment:
Does it mean that you have different number of blocks per archive (different number of examples per archive)?

Reply:
Yes, we can potentially have a different number of blocks per each archive. For e.g. if the actual blocks_per_archive is 72.2 then most of the time `num_blocks_per_this_archive` will be 72 but sometime it will be 73.